ARCs/arc-0037.md

```diff
@@ -0,0 +1,77 @@
+---
+arc: 37
+title: ABI Versioning
+description: Add a versioning scheme to the ABI
+author: Joe Polny (@joe-p)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/210
+status: Draft
+type: Standards Track
+category: ARC
+created: 2023-07-15
+requires: 4
+---
+
+## Abstract
+The Algorand ABI is defined in [ARC-0004](./arc-0004.md). As the protocol and desires of the ABI changes, there will be changes to the ABI. As such, there should be a versioning number to make it clear what versions of the ABI smart contracts adhere to and what versions of the ABI off-chain code supports.
+
+## Motivation
+Without any versioning scheme, it will be very hard to follow the evolution of the Algorand ABI.
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+Every ABI change is versioned `MAJOR.MINOR`. `MAJOR` is incremeneted when there are breaking changes and `MINOR` is incremented for the addition of backwards-compatible features.
```

Comment:
Some general questions about this scheme:
* Who decides what constitutes a new version?
* This version scheme is inherently linear. Why do you favor this approach instead of a more piecemeal one, e.g. where a contract lists the additional ARCs that it wants to use beyond ARC-4?

Reply:
> Who decides what constitutes a new version?

Any ARC that changes the ABI would have to include the proposed new version. 

> This version scheme is inherently linear. Why do you favor this approach instead of a more piecemeal one, e.g. where a contract lists the additional ARCs that it wants to use beyond ARC-4?

The main reason is simplicity. Changes to the ABI can be both additive and reductive so it can be hard to parse exactly what features support. For example, saying a contract is ARC-0004 compliant now becomes meaningless because you have to supply every single ABI-related ARC that it uses. Saying ABI 1.0 or ABI 2.0 is much more clear. It also makes it clear for when clients are backwards incompatible.

Reply:
> Any ARC that changes the ABI would have to include the proposed new version.

The order in which ARCs are proposed is not necessarily the order in which they will be finalized. Maybe ARCs could suggest what type of update they would be (i.e. major vs minor), then a maintainer can assign an actual version at the time of finalization?

Regardless of the decision here, I think this process should definitely be mentioned in this ARC.

Reply:
Yeah very good point. I will definitely add the process to this ARC and perhaps it could be a living ARC that contains a list of version changes and their respective ARCs